File: js/protocol/ledger.js
// Recomputation of a publisher's hash chain, entry balance and anchor commitment.
//
// Every rule here is stated at https://rgbmap.org/docs/ledgers-and-anchors; the
// publisher's implementations are
// backend trade-api-rs/src/ledger.rs (entry hash) and rgb-gateway-rs/src/snapshot.rs
// (anchor commitment).
//
// 🚨 The chain hashes commitments, not contents. A record's public half is published only for
// the platform events the publisher opens in full; for everything else the feed carries a
// commitment and nothing more, and the chain still recomputes end to end.
//
// 🚨 Where `public` is present, hash it verbatim: parsing and re-serialising it may change the
// byte sequence and produces a different digest.

const SEP = '\x1f';
const enc = new TextEncoder();

export const GENESIS_HASH = '0'.repeat(64);

export const hex = (bytes) => [...new Uint8Array(bytes)].map((b) => b.toString(16).padStart(2, '0')).join('');

export async function sha256Hex(input) {
    const bytes = typeof input === 'string' ? enc.encode(input) : input;
    return hex(await crypto.subtle.digest('SHA-256', bytes));
}

/** The hash of one ledger entry. Fields join with a single 0x1F byte. */
export function entryHash(entry) {
    return sha256Hex([entry.prevHash, entry.seq, entry.ts, entry.publicCommitment, entry.privateCommitment].join(SEP));
}

/**
 * The commitment to a record's public half.
 *
 * Only records the publisher opens in full carry `public` and `publicSalt`; for those, this
 * recomputes what the chain committed to, so the bytes can be checked rather than believed.
 */
export function publicCommitment(publicBytes, publicSalt) {
    return sha256Hex(['rgb-ledger-pub-v1', publicBytes, publicSalt].join(SEP));
}

/** The anchor commitment a snapshot claims, `rgb-anchor-v1`. */
export function anchorCommitment(snapshot) {
    const liabilities = Object.entries(snapshot.liabilities || {})
        .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
        .map(([asset, amount]) => `${asset}=${amount}`)
        .join(',');
    return sha256Hex([
        'rgb-anchor-v1',
        snapshot.network,
        String(snapshot.snapshotId),
        String(snapshot.ledgerSeq),
        snapshot.ledgerHash,
        snapshot.liabilitiesRoot,
        liabilities,
    ].join(SEP));
}

/** The OP_RETURN scriptPubKey hex that carries a 32-byte commitment. */
export const opReturnFor = (commitment) => '6a20' + commitment;

/**
 * Whether an entry's legs sum to zero per asset.
 *
 * Entries without legs (`issue`, `param_change`, …) are balanced by definition.
 */
export function legsBalance(publicJson) {
    const sums = new Map();
    for (const leg of publicJson.legs || []) {
        sums.set(leg.assetId, (sums.get(leg.assetId) || 0n) + BigInt(leg.delta));
    }
    return [...sums.values()].every((v) => v === 0n);
}

/**
 * Recompute a run of entries.
 *
 * `prevHash` is the hash the run continues from: the genesis constant for a run that
 * starts at seq 1. Returns the problems found; an empty array means the run recomputes.
 */
export async function verifyEntries(entries, prevHash = GENESIS_HASH, expectedFirstSeq = 1) {
    const problems = [];
    let prev = prevHash;
    let expected = BigInt(expectedFirstSeq);
    for (const e of entries) {
        if (BigInt(e.seq) !== expected) problems.push(`record #${e.seq}: expected seq ${expected}`);
        if (e.prevHash !== prev) problems.push(`record #${e.seq}: prevHash does not continue the chain`);
        if ((await entryHash(e)) !== e.hash) problems.push(`record #${e.seq}: hash does not match its contents`);

        // A record opened in full has to match the commitment the chain carries. A record
        // that is not opened is not a problem: the chain says nothing about its contents.
        if (e.public != null && e.publicSalt != null) {
            if ((await publicCommitment(e.public, e.publicSalt)) !== e.publicCommitment) {
                problems.push(`record #${e.seq}: public bytes do not match the commitment`);
            }
            let parsed;
            try {
                parsed = JSON.parse(e.public);
            } catch {
                problems.push(`record #${e.seq}: public segment is not JSON`);
                parsed = {};
            }
            if (!legsBalance(parsed)) problems.push(`record #${e.seq}: legs do not sum to zero`);
        }

        prev = e.hash;
        expected += 1n;
    }
    return problems;
}

/** Fields of the public segment that would mean private data leaked into it. */
export const PRIVATE_FIELDS = ['userId', 'address', 'destination', 'recipientId', 'vout'];

export function leakedFields(publicJson) {
    return PRIVATE_FIELDS.filter((k) => k in publicJson);
}

File: rust/rgbmap/src/ledger.rs
//! The ledger layer: the hash chain, the commitments and the balance rule.
//!
//! Specified at <https://rgbmap.org/docs/ledgers-and-anchors>.

use std::collections::BTreeMap;

use serde_json::Value;
use sha2::{Digest, Sha256};

/// Fields of a hashed tuple are joined by one 0x1F byte.
pub const SEP: u8 = 0x1F;

/// The `prev_hash` the first record continues from.
pub const GENESIS: &str = "0000000000000000000000000000000000000000000000000000000000000000";

pub fn sha256_hex(bytes: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    hasher.finalize().iter().map(|b| format!("{b:02x}")).collect()
}

/// The entry hash: fields joined by a single 0x1F byte, hashed as UTF-8.
///
/// The chain commits to the public half rather than carrying it, so a record hashes the same
/// whether or not its contents were published.
pub fn entry_hash(prev: &str, seq: &str, ts: &str, public_commitment: &str, private_commitment: &str) -> String {
    let mut buf = Vec::new();
    for (i, part) in [prev, seq, ts, public_commitment, private_commitment].iter().enumerate() {
        if i > 0 {
            buf.push(SEP);
        }
        buf.extend_from_slice(part.as_bytes());
    }
    sha256_hex(&buf)
}

/// The commitment to a record's public half, for the records published in full.
pub fn public_commitment(public: &str, salt: &str) -> String {
    let mut buf = Vec::new();
    for (i, part) in ["rgb-ledger-pub-v1", public, salt].iter().enumerate() {
        if i > 0 {
            buf.push(SEP);
        }
        buf.extend_from_slice(part.as_bytes());
    }
    sha256_hex(&buf)
}

/// The anchor commitment a snapshot claims.
pub fn anchor_commitment(snapshot: &Value) -> String {
    let liabilities = snapshot
        .get("liabilities")
        .and_then(Value::as_object)
        .map(|map| {
            map.iter()
                .map(|(asset, amount)| (asset.clone(), string_of(amount)))
                .collect::<BTreeMap<_, _>>()
                .iter()
                .map(|(asset, amount)| format!("{asset}={amount}"))
                .collect::<Vec<_>>()
                .join(",")
        })
        .unwrap_or_default();

    let parts = [
        "rgb-anchor-v1".to_string(),
        string_of(&snapshot["network"]),
        string_of(&snapshot["snapshotId"]),
        string_of(&snapshot["ledgerSeq"]),
        string_of(&snapshot["ledgerHash"]),
        string_of(&snapshot["liabilitiesRoot"]),
        liabilities,
    ];

    let mut buf = Vec::new();
    for (i, part) in parts.iter().enumerate() {
        if i > 0 {
            buf.push(SEP);
        }
        buf.extend_from_slice(part.as_bytes());
    }
    sha256_hex(&buf)
}

/// Numbers arrive as strings or as JSON numbers depending on the field; both hash as decimal.
pub fn string_of(value: &Value) -> String {
    match value {
        Value::String(s) => s.clone(),
        Value::Null => String::new(),
        other => other.to_string(),
    }
}

/// Whether one record's legs sum to zero per asset. i128 covers sats and token amounts.
pub fn legs_balance(public: &Value) -> bool {
    let Some(legs) = public.get("legs").and_then(Value::as_array) else {
        return true;
    };
    let mut sums: BTreeMap<String, i128> = BTreeMap::new();
    for leg in legs {
        let asset = string_of(&leg["assetId"]);
        let Ok(delta) = string_of(&leg["delta"]).parse::<i128>() else {
            return false;
        };
        *sums.entry(asset).or_insert(0) += delta;
    }
    sums.values().all(|v| *v == 0)
}
